crypt: add option to not encrypt data #1077

Vesnyx opened the pull request:
#### What is the purpose of this change?

As discussed in #1077, I'm adding an option to leave the data unencrypted. This permits a user to have only the filenames encrypted if they so choose.

#### Was the change discussed in an issue or in the forum before?

#1077
https://forum.rclone.org/t/is-it-possible-to-encrpyt-filenames-only/902

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

Vesnyx commented on backend/crypt/crypt.go:
```diff
@@ -329,10 +344,15 @@ type putFn func(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.O
+	wrappedIn, err := in, a
+
+	if f.opt.DataEncryption {
+			// Encrypt the data into wrappedIn
```
Not familiar with go, so I'm assuming this is some sort of error in the formatting?

Vesnyx commented:
I don't know how to fix the ineffectual assignment errors. It compiles fine for me.

ncw commented:
It is complaining about you assigning to err, then never using the result.

Just write

    var err error

Instead of assigning error(nil)

ncw commented:
This is looking good now :-)

Can you add a test?

This should be similar to this one

https://github.com/ncw/rclone/blob/master/backend/crypt/crypt_test.go#L68

But set the encryption to be off.  This will tell us whether everything is working.

Vesnyx commented:
Thanks for putting up with my incredibly coarse Go code. :) I've added the tests and also fixed the errors from when I was testing it earlier. I had forgotten to add the exceptions for the file seeks. 

Anyway, everything should be working now. If everything else looks good, I'l start working on the documentation.

Vesnyx commented:
@ncw It seems like the docs auto-generate. Do I need to add anything to https://github.com/ncw/rclone/blob/master/docs/content/crypt.md or is this all that's necessary?

ncw commented:
> It seems like the docs auto-generate. Do I need to add anything to https://github.com/ncw/rclone/blob/master/docs/content/crypt.md or is this all that's necessary?

The docs get auto generated on a release so that is fine.

Apologies for the delay, there has been something niggling me about this and I've finally worked out what!

I think we should invert the sense of `data_encryption` and call it `no_data_encryption`.  Then we can set the `Advanced: true` in the config as I think this is a valid but uncommon requirement.

What do you think of making those changes?

Vesnyx commented:
Yes, that actually makes a lot more sense. I didn't even realize there was an `Advanced: true` option till you brought it up. I'll refactor it following exam season in a few days. :)

ncw commented:
> Yes, that actually makes a lot more sense. I didn't even realize there was an `Advanced: true` option till you brought it up. I'll refactor it following exam season in a few days. :)

Thank you and good luck with your exams :-)

ncw commented:
How are you doing @Vesnyx - have you got time to continue with this?

Vesnyx commented:
Alright, sorry for the huge delay, been super busy recently. I'll be pushing the changes up as soon as I get back to my place.

ncw commented:
> Alright, sorry for the huge delay, been super busy recently. I'll be pushing the changes up as soon as I get back to my place.

No worries.. Ping me here when you are done :-)

Vesnyx commented:
Hey @ncw , I think I've finished up the majority of the changes. I also tried testing it locally and it appears to work and pass the tests. Let me know what else I need to do before it's ready to be added.

horganizzato commented:
is this still being considered?

ncw commented:
Sorry I've dropped the ball on this PR.

It needs rebasing against master and gofmt running on it, but it looks basically ready otherwise.

ivandeex commented:
@Vesnyx Do you need a hand rebasing?

Max-Sum commented:
Can I help rebasing this?

ncw commented:
@Max-Sum wrote:
> Can I help rebasing this?

Yes please - that would be fantastic.

Vesnyx commented:
Sorry everyone. Been swamped with work and haven't had a chance to work on this. I've just pushed the rebased commits.

ivandeex commented:
@Vesnyx 

I've pushed a commit addressing build errors.

```
golangci-lint run ./backend/crypt/...
backend/crypt/crypt.go:918:9: `if` block ends with a `return` statement, so drop this `else` and outdent its block (golint)
```
Just do what it wants.

`File is not goimportsed`
Just run `gofmt -w ./backend/crypt/crypt.go`

Please run this to obtain my commit: `cd /path/to/rclone; git fetch -v; git pull --ff-only`

Vesnyx commented:
It has been done. :)

ivandeex commented:
Rebased again due to recent breaking changes, to be sure.

ivandeex commented:
@ncw please review

ivandeex commented:
@ncw ping

ncw commented on backend/crypt/crypt.go:
```diff
@@ -351,6 +367,9 @@ func (f *Fs) put(ctx context.Context, in io.Reader, src fs.ObjectInfo, options [
 	if err != nil {
 		return nil, err
 	}
+	if f.opt.NoDataEncryption {
```
This is doing a lot more work than it needs to.

I think if you put this at the start of the function then it should work

```go
if f.opt.NoDataEncryption {
    return put(ctx, wrappedIn, f.newObjectInfo(src, nonce{}), options...)
}
```

ncw reviewed (commented):
Sorry for the delay in reviewing.

There are quite a few opportunities for making this simpler and more efficient which I've put in comments inline.

I think the basic idea is sound and the tests are good.

ivandeex commented:
@Vesnyx please address review comments

ivandeex commented:
Rebased on current master

ivandeex commented on backend/crypt/crypt.go:
```diff
@@ -855,6 +901,24 @@ func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (rc io.Read
 		}
 	}
+
+	if o.f.opt.NoDataEncryption {
```
done

ivandeex commented on backend/crypt/crypt.go:
```diff
@@ -822,9 +863,13 @@ func (o *Object) Remote() string {
+	size := o.Object.Size()
+	if !o.f.opt.NoDataEncryption {
+		var err error
+		size, err = o.f.cipher.DecryptedSize(o.Object.Size())
```
done

ivandeex commented on backend/crypt/crypt.go:
```diff
@@ -617,12 +636,34 @@ func (f *Fs) computeHashWithNonce(ctx context.Context, nonce nonce, src fs.Objec
 // Note that we break lots of encapsulation in this function.
 func (f *Fs) ComputeHash(ctx context.Context, o *Object, src fs.Object, hashType hash.Type) (hashStr string, err error) {
+	if f.opt.NoDataEncryption {
+		// Open the src for input
```
done

ivandeex commented:
@ncw I have addressed your review comments. Please this PR. Thank you.

ivandeex commented:
@ncw
I will merge this myself.
Please take care of the rest.

ncw commented:
Thank you :-)
